test/net/sourceforge/kolmafia/request/RequestTestBase.java

```diff
@@ -32,4 +86,171 @@ protected void expectSuccess(GenericRequest spy, String responseText) {
+   */
+  protected ParamMatcher hasParam(final String param, final String value) {
+    return new ParamMatcher(param, value);
+  }
+
+  /**
+   * Is true if one of the URL's parameters has a key matching {@code param}, and a value matching
+   * {@code value}. The parameters are the part coming after the {@code ?} (if any), separated by
+   * {@code &}s.
+   *
+   * <p>For example, in {@code inventory.php?which=2&foo&bar=&pwd=abc}, the parameters are {@code
+   * which=2}, {@code foo}, {@code bar=} and {@code pwd=abc}.
+   *
+   * <p>Parameters are split into a key and a value, the key being what comes before the {@code =}
+   * (if any), and the value being what's after (again, if any).
+   */
+  protected ParamMatcher hasParam(final String param, final Pattern valueMatcher) {
+    return new ParamMatcher(param, valueMatcher);
+  }
+
+  private void prepareResponse(final String response, final ArgumentMatcher<URL> matcher) {
+    try {
+      when(factory.openConnection(argThat(matcher))).thenReturn(new FakeConnection(response));
+    } catch (IOException e) {
+    }
+  }
+
+  private class FakeConnection extends HttpURLConnection {
+    private final String responseText;
+
+    private FakeConnection(final String responseText) {
+      super(null);
+      this.responseText = responseText;
+      this.responseCode = 200;
+    }
+
+    @Override
+    public InputStream getInputStream() throws IOException {
+      return new ByteArrayInputStream(this.responseText.getBytes(StandardCharsets.UTF_8));
+    }
+
+    @Override
+    public void disconnect() {}
+
+    @Override
+    public boolean usingProxy() {
+      return false;
+    }
+
+    @Override
+    public void connect() throws IOException {}
+  }
+
+  private class ParsedUrl {
+    private final String path;
+    private final Map<String, String> params = new HashMap<>();
+
+    private ParsedUrl(final URL url) {
+      this.path = url.getPath();
+      final String query = url.getQuery();
+
+      if (query != null) {
+        final String[] paramsString = query.split("&");
+        for (final String param : paramsString) {
+          if (param.isEmpty()) {
+            continue;
+          }
+
+          final String[] splitParam = param.split("=", 2);
+          params.put(splitParam[0], splitParam.length > 1 ? splitParam[1] : null);
+        }
+      }
+    }
+  }
+
+  private abstract class UrlMatcher {
+    abstract boolean matches(final ParsedUrl url);
+  }
+
+  private class PathMatcher extends UrlMatcher {
+    final ArgumentMatcher<String> matcher;
+
+    private PathMatcher(final String exactPath) {
+      this.matcher = path -> path.equals(exactPath);
+    }
+
+    boolean matches(final ParsedUrl url) {
+      return this.matcher.matches(url.path);
+    }
+  }
+
+  private class ParamMatcher extends UrlMatcher {
+    final ArgumentMatcher<Entry<String, String>> matcher;
+
+    private ParamMatcher(final String param, final String value) {
+      this.matcher =
```

Comment:
Why not store the String param and an ArgumentMatcher\<String> matcher?

That way instead of iterating through all the entries of params, you can use something like `matcher.matches(params.get(param))`.

Reply:
I think I was still thinking about users maybe wanting to also submit a Pattern for the key.
I changed my mind mid-way, as the keys are pretty much specific and predictable enough that the user will know what to expect.